File: python/rapidsmpf/rapidsmpf/tests/test_shuffler_packed_data.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES.
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from typing import TYPE_CHECKING

from rapidsmpf.memory.buffer_resource import BufferResource
from rapidsmpf.memory.packed_data import PackedData
from rapidsmpf.shuffler import Shuffler

if TYPE_CHECKING:
    import rmm

    from rapidsmpf.communicator.communicator import Communicator


def test_shuffler_packed_data_roundtrip(
    comm: Communicator,
    device_mr: rmm.mr.CudaMemoryResource,
) -> None:
    br = BufferResource(device_mr)
    packed = PackedData.from_host_bytes(b"hello world", br)
    shuffler = Shuffler(comm, op_id=0, total_num_partitions=1, br=br)

    try:
        shuffler.insert_chunks({0: packed})
        shuffler.insert_finished()
        shuffler.wait()

        result = shuffler.extract(0)
        assert len(result) == 1
        assert result[0].to_host_bytes() == b"hello world"
    finally:
        shuffler.shutdown()

File: cpp/tests/test_shuffler_packed_data.cpp
/**
 * SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <chrono>
#include <unordered_map>

#include <gtest/gtest.h>

#include <rmm/mr/cuda_memory_resource.hpp>

#include <rapidsmpf/communicator/single.hpp>
#include <rapidsmpf/memory/packed_data.hpp>
#include <rapidsmpf/shuffler/shuffler.hpp>

#include "environment.hpp"
#include "utils.hpp"

extern Environment* GlobalEnvironment;

TEST(ShufflerPackedData, basic_roundtrip) {
    auto const& comm = GlobalEnvironment->comm_;
    auto br = rapidsmpf::BufferResource::create(rmm::mr::cuda_memory_resource{});
    auto stream = br->stream_pool().get_stream();

    rapidsmpf::shuffler::Shuffler shuffler{
        comm, 0, 1, br.get(), rapidsmpf::shuffler::Shuffler::round_robin
    };

    std::unordered_map<rapidsmpf::shuffler::PartID, rapidsmpf::PackedData> chunks;
    chunks.emplace(0, generate_packed_data(4, 0, stream, *br));
    shuffler.insert(std::move(chunks));
    shuffler.insert_finished();
    shuffler.wait(std::chrono::seconds{30});

    auto results = shuffler.extract(0);
    EXPECT_FALSE(results.empty());
}
